aten/src/ATen/cudnn/Descriptors.h

```diff
@@ -41,20 +41,47 @@ static inline void fixSizeOneDimStride(int dim, const int *size, int *stride) {
   }
 }
 
-// TODO: Use unique_ptr on this?
-struct TensorDescriptor
-{
-  cudnnTensorDescriptor_t desc;
+template <typename T, cudnnStatus_t (*dtor)(T*)>
+struct DescriptorDeleter {
+  void operator()(T* x) {
+    CUDNN_CHECK(dtor(x));
+  }
+};
 
-  explicit TensorDescriptor() : desc(nullptr) {
-    CUDNN_CHECK(cudnnCreateTensorDescriptor(&desc));
-  };
+// A generic class for wrapping cuDNN descriptor types.  All you need
+// is to give the underlying type the Descriptor_t points to (usually,
+// if it's cudnnTensorDescriptor_t it points to cudnnTensorStruct),
+// the constructor and the destructor.  Subclasses are responsible
+// for forwarding constructors and defining a set() function to actually
+// set the descriptor.
+template <typename T, cudnnStatus_t (*ctor)(T**), cudnnStatus_t (*dtor)(T*)>
+class Descriptor
+{
+public:
+  explicit Descriptor() {
+    T* raw_desc;
+    CUDNN_CHECK(ctor(&raw_desc));
+    desc_.reset(raw_desc);
+  }
 
-  TensorDescriptor(const TensorDescriptor& ref) = delete;
-  TensorDescriptor(TensorDescriptor&& ref) = delete;
+  // TODO: Figure out why const-correctness doesn't work here
+  T* desc() const { return desc_.get(); }
+  T* desc() { return desc_.get(); }
+private:
+  std::unique_ptr<T, DescriptorDeleter<T, dtor>> desc_;
+};
 
-  ~TensorDescriptor() {
-    cudnnDestroyTensorDescriptor(desc);
+class TensorDescriptor
+  : public Descriptor<cudnnTensorStruct,
```

Comment:
Why do we use `cudnnTensorStruct` instead of `cudnnTensorDescriptor_t`? Is this even part of the public API?

Reply:
So that there is not an extra level of indirection ;)

It's in the header file, so I'm assuming it's part of the public API. @ngimel, you want to weigh in?

Reply:
It is indeed in the header file (and they should not have made it opaque in the first place, if you ask me).